src/cmd/metadata/contract.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::fmt::{Display, Formatter, Result as DisplayResult, Write};
+use semver::Version;
+use serde::{Serialize, Serializer};
+use serde_json::{Map, Value};
+use url::Url;
+
+const METADATA_VERSION: &str = "0.1.0";
+
+/// An entire ink! project for metadata file generation purposes.
+#[derive(Debug, Serialize)]
+pub struct ContractMetadata {
+    metadata_version: semver::Version,
+    source: Source,
+    contract: Contract,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    user: Option<User>,
+    /// Raw JSON of the metadata generated by the ink! contract itself
+    #[serde(flatten)]
+    ink: Map<String, Value>,
```

Comment:
Not optimal this this is not really checked or intransparent to `cargo-contract` but I guess this is to not create some dependency on to `ink_metadata` crate?

Reply:
Yes exactly that. It's an evolution of the previous approach, which was just writing the serialized metadata produced by `ink!` metadata to a file. It maintains the behaviour of forwards/backwards compatibility with different versions of `ink!`